pymaker/__init__.py

```diff
@@ -681,6 +734,74 @@ def invocation(self) -> Invocation:
         return Invocation(self.address, Calldata(self._contract_function()._encode_transaction_data()))
 
 
+class RecoveredTransact(Transact):
+    """ Models a pending transaction retrieved from the mempool.
+
+    These can be created by a call to `get_pending_transactions`, enabling the consumer to implement logic which
+    cancels pending transactions upon keeper/bot startup.
+    """
+    def __init__(self, web3: Web3,
+                 address: Address,
+                 nonce: int,
+                 latest_tx_hash: str,
+                 current_gas: int):
+        assert isinstance(current_gas, int)
+        super().__init__(origin=None,
+                         web3=web3,
+                         abi=None,
+                         address=address,
+                         contract=None,
+                         function_name=None,
+                         parameters=None)
+        self.nonce = nonce
+        self.tx_hashes.append(latest_tx_hash)
+        self.current_gas = current_gas
+
+    def name(self):
+        return f"Recovered tx with nonce {self.nonce}"
+
+    @_track_status
+    async def transact_async(self, **kwargs) -> Optional[Receipt]:
+        # TODO: Read transaction data from chain, create a new state machine to manage gas for the transaction.
+        raise NotImplementedError()
+
+    def cancel(self, gas_price: GasPrice):
+        return synchronize([self.cancel_async(gas_price)])[0]
+
+    async def cancel_async(self, gas_price: GasPrice):
+        assert isinstance(gas_price, GasPrice)
+        initial_time = time.time()
+        self.gas_price_last = self.current_gas
+        self.tx_hashes.clear()
+
+        if gas_price.get_gas_price(0) <= self.current_gas * 1.125:
```

Comment:
This multiplier seems high, even a minuscule increase in the price should be enough to override the previous transaction as long as it is broadcast to the miner. I wonder if this could be a default value with an optional configurable multiplier. Foundation keepers may optimize for speed, but I can imagine strategies that might want to cut it a little closer.

Reply:
Counter to one's intuition, each client has a minimum % increase threshold needed to replace a transaction with the same nonce. It's 12.5% for [OpenEthereum](https://openethereum.github.io/wiki/Transactions-Queue) and (IIRC) 10% for geth. Therefore, I believe surpassing the 12.5% threshold was the intention.

Reply:
12.5% is the minimum increase required to be accepted by a Parity/OpenEthereum node (refer to [Dropping conditions](https://openethereum.github.io/wiki/Transactions-Queue) section).  And yes, the `gas_price` passed by the consumer can be configured as desired.